src/cider/exec/nextgen/function/CiderStringFunctions.cpp

```diff
@@ -154,3 +156,139 @@ extern "C" RUNTIME_EXPORT int64_t cider_ascii_upper(int8_t* string_heap_ptr,
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->addString(str.data(), str.length());
+  return pack_string((const int8_t*)s.getDataUnsafe(), (const int32_t)s.getSize());
+}
+
+extern "C" RUNTIME_EXPORT NEVER_INLINE int64_t
+gen_string_from_double(const double operand, char* string_heap_ptr) {
+  std::string str = fmt::format("{:#}", operand);
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->addString(str.data(), str.length());
+  return pack_string((const int8_t*)s.getDataUnsafe(), (const int32_t)s.getSize());
+}
+
+extern "C" RUNTIME_EXPORT ALWAYS_INLINE int64_t
+gen_string_from_bool(const int8_t operand, char* string_heap_ptr) {
+  std::string str = (operand == 1) ? "true" : "false";
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->addString(str.data(), str.length());
+  return pack_string((const int8_t*)s.getDataUnsafe(), (const int32_t)s.getSize());
+}
+
+extern "C" RUNTIME_EXPORT ALWAYS_INLINE int64_t
+gen_string_from_time(const int64_t operand,
+                     char* string_heap_ptr,
+                     const int32_t dimension) {
+  constexpr size_t buf_size = 64;
+  char buf[buf_size];
+  int32_t str_len = shared::formatHMS(buf, buf_size, operand, dimension);
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->addString(buf, str_len);
+  return pack_string((const int8_t*)s.getDataUnsafe(), (const int32_t)s.getSize());
+}
+
+extern "C" RUNTIME_EXPORT ALWAYS_INLINE int64_t
+gen_string_from_timestamp(const int64_t operand,
+                          char* string_heap_ptr,
+                          const int32_t dimension) {
+  constexpr size_t buf_size = 64;
+  char buf[buf_size];
+  int32_t str_len = shared::formatDateTime(buf, buf_size, operand, dimension);
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->addString(buf, str_len);
+  return pack_string((const int8_t*)s.getDataUnsafe(), (const int32_t)s.getSize());
+}
+
+extern "C" RUNTIME_EXPORT ALWAYS_INLINE int64_t
+gen_string_from_date(const int32_t operand, char* string_heap_ptr) {
+  constexpr size_t buf_size = 64;
+  char buf[buf_size];
+  int32_t str_len = shared::formatDays(buf, buf_size, operand);
+  StringHeap* ptr = reinterpret_cast<StringHeap*>(string_heap_ptr);
+  string_t s = ptr->addString(buf, str_len);
+  return pack_string((const int8_t*)s.getDataUnsafe(), (const int32_t)s.getSize());
+}
+
+#define DEF_CONVERT_STRING_TO_INTEGER(value_type, value_name)          \
+  extern "C" RUNTIME_EXPORT value_type convert_string_to_##value_name( \
+      const char* str_ptr, const int32_t str_len) {                    \
+    std::string from_str(str_ptr, str_len);                            \
+    value_type res = std::stoi(from_str);                              \
+    if (res >= std::numeric_limits<value_type>::min() &&               \
```

Comment:
The res is ```value_type```, should it always be in ```value_type```'s range?

Reply:
maybe need to add a try-catch around stoi

Reply:
yes, stoi will return a value of integer type, need value limits check if value type is int16_t/int8_t.

Reply:
and just throw runtime exception.

Reply:
I think stoi's runtime error would be better wrapped in cider-runtime error